src/common/Gateway.sol

```diff
@@ -285,86 +149,46 @@ contract Gateway is Auth, Recoverable, IGateway {
         internal
         returns (bool succeeded)
     {
-        IAdapter[] memory adapters_ = adapters[centrifugeId];
-        require(adapters[centrifugeId].length != 0, EmptyAdapterSet());
-
-        SendData memory data = SendData({
-            batchHash: keccak256(batch),
-            payloadId: bytes32(""),
-            gasCost: new uint256[](MAX_ADAPTER_COUNT)
-        });
-        data.payloadId = keccak256(abi.encodePacked(localCentrifugeId, centrifugeId, data.batchHash));
-
-        {
-            // Estimate gas usage
-            uint256 total;
-            for (uint256 i; i < adapters_.length; i++) {
-                data.gasCost[i] = adapters_[i].estimate(
-                    centrifugeId,
-                    i == PRIMARY_ADAPTER_ID - 1 ? batch : data.batchHash.serializeMessageProof(),
-                    batchGasLimit
-                );
-
-                total += data.gasCost[i];
+        bytes32 batchHash = keccak256(batch);
+        uint256 cost = adapter.estimate(centrifugeId, batch, batchGasLimit);
+
+        // Ensure sufficient funds are available
+        if (transactionRefund != address(0)) {
+            require(cost <= fuel, NotEnoughTransactionGas());
+            fuel -= cost;
+        } else { // Subsidized pool payment
+            if (cost > subsidy[poolId].value) {
+                _requestPoolFunding(poolId);
             }
 
-            // Ensure sufficient funds are available
-            if (transactionRefund != address(0)) {
-                require(total <= fuel, NotEnoughTransactionGas());
-                fuel -= total;
+            if (cost <= subsidy[poolId].value) {
+                subsidy[poolId].value -= uint96(cost);
             } else {
-                if (total > subsidy[poolId].value) {
-                    _requestPoolFunding(poolId);
-                }
-
-                if (total <= subsidy[poolId].value) {
-                    subsidy[poolId].value -= uint96(total);
-                } else {
-                    underpaid[centrifugeId][data.batchHash].counter++;
-                    underpaid[centrifugeId][data.batchHash].gasLimit = batchGasLimit;
-                    emit UnderpaidBatch(centrifugeId, batch);
-                    return false;
-                }
-            }
-        }
+                Underpaid storage underpaid_ = underpaid[centrifugeId][batchHash];
```

Comment:
I think we could call `addUnpaidMessage()` here

Reply:
Super great recommendation!

Reply:
Ah, I now remember I couldn't not. The gasLimit computation differs.

Reply:
I did this anyway: [9ed8b25](https://github.com/centrifuge/protocol-v3/pull/372/commits/9ed8b25d93d16d4ab51215e0b88f1fe94c9b0eb0)